### ISSUE TITLE
When should the TEU schedule be written to the battery

### ISSUE THREAD

**Issue by ruteclrp** [billerby/huawei-solar-ha-nodered#8]
The script fetches the new tomorrow's prices from Nordpool @ 13:05 on day 1.
With the current setup it runs the code and write the schedule for day 2 to the battery at around 13;30-14:00 depending on when the new prices were becoming available.
The issue is tha the TOU schedule for day 1 only ends at 23:59 on day 1 - several hours after the script attempts to write the day 2 schedule.
This creates a conflict.
The script should not write the day 2 schedule until end of day 1 - i.e. around 23:59 of day 1

Looking forward to hear you comments

**Comment by billerby** [comment]
Hi!

In the periods that written to the schedule I also specify the day. So writing a period for Tuesday should not affect Monday. BTW, have you updated to the latest version? I fixed several bugs after I initially posted the flow. For example running my script twice in one day, does not work in the current setup I think (writing the same day twice), so you have to clear the periods first (I do that calling the service directly in developer tools (yml).

For me to find out whats wrong please post the logs from the node thats failing.

With that said, I posted my flow in a swedish fusionsolar group and another user showed up with his own node-red node that he implemented for our batteries. He implemented a genetic algorithm that is rather complex which schedules the battery in minutes in a real optimized way, I am now testing that instead. 

With his node my flow now looks this tiny and manageble :)
![image](https://user-images.githubusercontent.com/123237/211138552-737b76af-fa56-40a7-836a-9b64756b2efd.png)

You can find his node in the palette, node-red-contrib-genetic-charging-strategy

https://github.com/enell/node-red-contrib-genetic-charging-scheduler

You configure it with you battery capacity, max input power, average consumption, and also the current state of charge which is read from the battery and injected to the node.

**Comment by ruteclrp** [comment]
Yes, you are right - did not think about that. Also think the error on writing the new schedule was due to existing everyday schedule in the battery which maybe could not be overwritten?
Took a look at the enel code. A bit confused as that contained much more than just a node-red .json file with the script. What should I do to use the enel package? the readme file is rather quickly red!

**Comment by billerby** [comment]
You can install the node type from  "manage palette".  
![image](https://user-images.githubusercontent.com/123237/211200818-dee89d83-c85b-4eb1-9e3e-f1eed25a49c7.png)

Then you connect a Price Receiver node from it, and also a current state from you battery via a join node:
![image](https://user-images.githubusercontent.com/123237/211200914-7a59cc92-a0c6-420f-a363-a94d71046b6b.png)

Here are the props you set on the scheduling node, I have kept the defaults for the genetic algorithm:
![image](https://user-images.githubusercontent.com/123237/211201082-4f98b20c-bf86-4cce-8c12-65f92ec86fa0.png)
Just input your battery capacity, the input power you are charging with and you average consumption per hour in your home.

Finally you need to convert the schedule from its output to the ToU-period for our battery in a function node, I have a gist with the code inside the function node here: https://gist.github.com/billerby/605dc8f5955fafb7f640900d0614b6df

![image](https://user-images.githubusercontent.com/123237/211201001-b61e3522-fc91-4cd8-8228-f6c8b82ff6cc.png)

The stuff below with the Optimized battery strategy is just a sensor that is used for an apex chart, an nothing you need for the scheduling.
